source/stun_deserializer.c

```diff
@@ -0,0 +1,868 @@
+#define STUN_WRITE_UINT64   ( pCtx->readWriteFn.WriteU64Fn )
+#define STUN_READ_UINT16    ( pCtx->readWriteFn.ReadU16Fn )
+#define STUN_READ_UINT32    ( pCtx->readWriteFn.ReadU32Fn )
+#define STUN_READ_UINT64    ( pCtx->readWriteFn.ReadU64Fn )
+
+/* Static Functions */
+static StunResult_t StunDeserializer_ParseAttributeUINT32( const StunContext_t * pCtx,
+                                                           const StunAttribute_t * pAttribute,
+                                                           uint32_t * val,
+                                                           StunAttributeType_t attributeType );
+
+static StunResult_t StunDeserializer_ParseAttributeUINT64( const StunContext_t * pCtx,
+                                                           const StunAttribute_t * pAttribute,
+                                                           uint64_t * val,
+                                                           StunAttributeType_t attributeType );
+
+static StunResult_t StunDeserializer_ParseAttributeBuffer( const StunAttribute_t * pAttribute,
+                                                           const char ** ppBuffer,
+                                                           uint16_t * pBufferLength,
+                                                           StunAttributeType_t attributeType );
+
+static void StunDeserializer_InitEndianness( StunContext_t * pCtx );
+
+/*-----------------------------------------------------------*/
+
+static StunResult_t StunDeserializer_ParseAttributeUINT32( const StunContext_t * pCtx,
+                                                           const StunAttribute_t * pAttribute,
+                                                           uint32_t * val,
+                                                           StunAttributeType_t attributeType )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
+    if( ( pAttribute == NULL ) ||
+        ( val == NULL ) ||
+        ( pAttribute->attributeType != attributeType ) ||
+        ( pAttribute->pAttributeValue == NULL ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        if( pAttribute->attributeValueLength != sizeof( uint32_t ) )
+        {
+            result = STUN_RESULT_INVALID_ATTRIBUTE_LENGTH;
+        }
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        STUN_READ_UINT32( val, ( uint8_t * ) &( *( ( uint32_t * ) pAttribute->pAttributeValue ) ) );
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+static StunResult_t StunDeserializer_ParseAttributeUINT64( const StunContext_t * pCtx,
+                                                           const StunAttribute_t * pAttribute,
+                                                           uint64_t * val,
+                                                           StunAttributeType_t attributeType )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
```

Comment:
pCtx is not used in this function, should it be passed as an argument?

Reply:
pCtx is used for STUN_READ_UINT64, which is used as this - #define STUN_READ_UINT64 ( pCtx->readWriteFn.ReadU64Fn )